{
 "metadata": {
  "name": "",
  "signature": "sha256:d37bb330c9d173a978fcc635dcb81892f2dc573557dc76f945db45cd9153167f"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Motif finding is more difficult than you think\n",
      "----\n",
      "\n"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import itertools\n",
      "import collections\n",
      "import MotMod\n",
      "\n",
      "def consume(iterator, n):\n",
      "    \"Advance the iterator n-steps ahead. If n is none, consume entirely.\"\n",
      "    # Use functions that consume iterators at C speed.\n",
      "    if n is None:\n",
      "        # feed the entire iterator into a zero-length deque\n",
      "        collections.deque(iterator, maxlen=0)\n",
      "    else:\n",
      "        # advance to the empty slice starting at position n\n",
      "        next(itertools.islice(iterator, n, n), None)\n",
      "    \n",
      "\n",
      "\n",
      "def find_Ham(stringone,stringtwo):\n",
      "    count = 0\n",
      "    for i,k in zip(stringone,stringtwo):\n",
      "        if i!=k:\n",
      "            count+=1\n",
      "    return count\n",
      "\n",
      "def stringpresent(Pattern,DNA,k,d):\n",
      "    for string in DNA:\n",
      "        for i in range(len(string)-k):\n",
      "            if find_Ham(string[i:i+k],Pattern)<=d:\n",
      "                return True\n",
      "    return False\n",
      "\n",
      "def Neighbors(Pattern,d):\n",
      "    if d==0:\n",
      "        return Pattern\n",
      "    bases = {0:\"A\",1:\"T\", 2:\"G\", 3:\"C\"}\n",
      "    if len(Pattern)==1:\n",
      "        return \"ACGT\"\n",
      "    Neighborhood = []\n",
      "    SuffixNeighbors = Neighbors(Pattern[1:len(Pattern)],d)\n",
      "    for i in SuffixNeighbors:\n",
      "        if find_Ham(Pattern[1:len(Pattern)],i) <= d:\n",
      "            for z in bases.values():\n",
      "                Neighborhood.append(z+i)\n",
      "        else:\n",
      "            Neighborhood.append(Pattern[0]+i)\n",
      "    return Neighborhood\n",
      "\n",
      "#Boolean returning function determines whether Pattern is present in every string of DNA with at most d-mismatches\n",
      "def Present_in_DNA(DNA,Pattern,k,d):\n",
      "    truths=0;i=0\n",
      "    for string in DNA: # for every string of DNA stored in DNA list\n",
      "        for i in range(len(string)-k): #for every nucleotide in the DNA string\n",
      "            if find_Ham(Pattern,string[i:i+k])<=d: # if there is a kmer in that dna string with <=d mismatches\n",
      "                consume(string,1) # move on to the next string rather than continuing to iterate through all the other patterns\n",
      "                truths+=1 # increase the counter detailing how often the kmer is found in the collective DNA set of strings\n",
      "    if truths>=len(DNA):\n",
      "        return True\n",
      "    else:\n",
      "        return False\n",
      "            \n",
      "\n",
      "def MotifEnumeration(DNA,k,d):\n",
      "    City=[]\n",
      "    Patterns=[string[i:i+k] for string in DNA for i in range(len(string)-k)]\n",
      "    City=[Neighbors(Pattern,d) for Pattern in Patterns] # collect all possible variations of kmers within d tolerance\n",
      "    BestPatterns=[Neighborhood[i] for Neighborhood in City for i in range(len(Neighborhood)) if Present_in_DNA(DNA,Neighborhood[i],k,d)]\n",
      "#     for Neighborhood in City: # for each list of possible kmers \n",
      "#         for Neighbor in Neighborhood: # for each possible kmer in each list\n",
      "#             if Present_in_DNA(DNA,Neighbor,k,d): # if that kmer is present in every string of DNA with tolerance level d\n",
      "#                 BestPatterns.append(Neighbor) # add that kmer to the \"Best Patterns\" list to return\n",
      "    BestPatterns=set(BestPatterns);BestPatterns=list(BestPatterns) # remove duplicates\n",
      "    return BestPatterns # return list\n",
      "\n",
      "\n",
      "Example=[\"ATTTGGC\",\"TGCCTTA\",\"CGGTATC\", \"GAAAATT\"]\n",
      "answers=[\"ATA\", \"ATT\", \"GTT\", \"TTT\"]\n",
      "\n",
      "# print Present_in_DNA(Example,\"TTT\",4,2)\n",
      "\n",
      "print \"\\n\".join(MotifEnumeration(Example,3,1))\n",
      "# print set(answers)==set(MotifEnumeration(Example,3,1))\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "CTT\n",
        "AAA\n",
        "TGT\n",
        "ATA\n",
        "GAT\n",
        "CAT\n",
        "AAT\n",
        "ATT\n",
        "GTT\n",
        "TAA\n",
        "TTT\n",
        "CGT\n",
        "GAA\n",
        "TGG\n"
       ]
      }
     ],
     "prompt_number": 20
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Scoring Motifs\n",
      "\n",
      "Use a matrix of the varying candidate kmers to create profile of similarity\n",
      "\n",
      "Four important concepts we'll use to build a better idea of what motif would most likely be found in as a TBF- all of these will be matrices of different types.\n",
      "\n",
      "1. Motif Matrix:\n",
      "    * Stores each potential motif kmer as a string delineated by individual nucleotides.\n",
      "    * dimensions: k x n where k is the length of the string, and n is the number of candidate motifs\n",
      "    * Use: This matrix can be used to understand which nucleotide is the most popular in each position of the kmer\n",
      "2. Count Matrix:\n",
      "    * Matrix holding the count of each nucleotide in each motif string COLUMN\n",
      "    * This is a 4 x k matrix where each row holds information pertaining to a specfic nucleotide\n",
      "    * Use: Used to construct the profile matrix\n",
      "3. Profile Matrix:\n",
      "    * Matrix holding the *frequency* of each Nucleotide in each column (i.e. position) of the nucleotide kmer\n",
      "    * dimensions: 4 x k again, where each row holds information pertaining to a specific nucleotide at each position in the kmer\n",
      "    * Use: Constructing the Consensus Motif\n",
      "4. Consensus Motif:\n",
      "    * Using the Profile matrix, we can determine what might be the most \"idealized\" candidate for a (k,d) motif\n",
      "    \n"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import numpy as np\n",
      "import math as m\n",
      "#Test data\n",
      "# s1=\"TCGGGGGTTTTT\"\n",
      "# s2=\"CCGGTGACTTAC\"\n",
      "# s3=\"ACGGGGATTTTC\"\n",
      "# s4=\"TTGGGGACTTTT\"\n",
      "# s5=\"AAGGGGACTTCC\"\n",
      "# s6=\"TTGGGGACTTCC\"\n",
      "# s7=\"TCGGGGATTCAT\"\n",
      "# s8=\"TCGGGGATTCCT\"\n",
      "# s9=\"TAGGGGAACTAC\"\n",
      "# s10=\"TCGGGTATAACC\"\n",
      "# sses=[s1,s2,s3,s4,s5,s6,s7,s8,s9,s10]\n",
      "\n",
      "def MotifMat(DNA):\n",
      "    return np.array([list(string) for string in DNA])\n",
      "\n",
      "def CountMat(MotifMatrix):\n",
      "    CountMatrix=np.empty([4,len(MotifMatrix[0])])\n",
      "    for i in range(len(MotifMatrix[0])):\n",
      "        CountMatrix[0,i]=sum(1 if x==\"A\" else 0 for x in MotifMatrix[:,i] )\n",
      "        CountMatrix[1,i]=sum(1 if x==\"C\" else 0 for x in MotifMatrix[:,i])\n",
      "        CountMatrix[2,i]=sum(1 if x==\"G\" else 0 for x in MotifMatrix[:,i])\n",
      "        CountMatrix[3,i]=sum(1 if x==\"T\" else 0 for x in MotifMatrix[:,i])\n",
      "    return CountMatrix\n",
      "\n",
      "def CountMatPseudo(MotifMatrix):\n",
      "    CountMatrix=np.empty([4,len(MotifMatrix[0])])\n",
      "    for i in range(len(MotifMatrix[0])):\n",
      "        CountMatrix[0,i]=sum(1 if x==\"A\" else 0 for x in MotifMatrix[:,i])+1\n",
      "        CountMatrix[1,i]=sum(1 if x==\"C\" else 0 for x in MotifMatrix[:,i])+1\n",
      "        CountMatrix[2,i]=sum(1 if x==\"G\" else 0 for x in MotifMatrix[:,i])+1\n",
      "        CountMatrix[3,i]=sum(1 if x==\"T\" else 0 for x in MotifMatrix[:,i])+1\n",
      "    return CountMatrix\n",
      "\n",
      "def ProfileMat(CountMatrix):\n",
      "    ProfileMatrix=np.empty([4,len(CountMatrix[0])])\n",
      "    for i in range(len(CountMatrix[0])):\n",
      "        ProfileMatrix[0,i]=(CountMatrix[0,i])/(sum(CountMatrix[:,i]))\n",
      "        ProfileMatrix[1,i]=CountMatrix[1,i]/sum(CountMatrix[:,i])\n",
      "        ProfileMatrix[2,i]=CountMatrix[2,i]/sum(CountMatrix[:,i])\n",
      "        ProfileMatrix[3,i]=CountMatrix[3,i]/sum(CountMatrix[:,i])\n",
      "    return ProfileMatrix\n",
      "\n",
      "def Get_Consensus(ProfileMatrix):\n",
      "    index = {0:\"A\", 1: \"C\", 2:\"G\", 3: \"T\"}\n",
      "    consensus_string=\"\"\n",
      "    for i in ProfileMatrix.argmax(axis=0):\n",
      "        consensus_string+=index[i]\n",
      "    return consensus_string\n",
      "\n",
      "def Get_Entropy(ProfileMatrix):\n",
      "    runningsum=[]\n",
      "    for i in range(len(ProfileMatrix[0])):\n",
      "        runningsum.append(Entropy_Calculation(ProfileMatrix[:,i]))\n",
      "    print runningsum\n",
      "    return sum(runningsum)\n",
      "        \n",
      "def Entropy_Calculation(l):\n",
      "    summation=0\n",
      "    for i in l:\n",
      "        if i==0:\n",
      "            pass\n",
      "        else:\n",
      "            summation+= i * m.log(i,2)\n",
      "    return -summation"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Let's keep thinking:\n",
      "We need to find an alternate way of computing the Score of motifs. To begin a better understanding of how we might do so, consider the fact that the \"Score\" could be as easily computed from a column x column summation, as it could have by a row by row summation (see image)\n",
      "<img src=\"img1_matrix.png\">\n",
      "\n",
      "This means something important- it means we can now define $d(Pattern, Motifs)$ as the sum of the Hamming Distances between Pattern and each Motif$_{f}$ $$ d(Pattern,Motifs) = \\sum_{i=1}^{t} HammingDistance(Pattern,Motif_{i})$$"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Eureka!\n",
      "\n",
      "So instead of searching for a collection of kmers *Motifs* by minimizing $Scores(Motifs)$ we can instead search for a potential consensus string *Pattern* minimizing $d(Pattern,Motifs)$ among all possible kmers and all possible choices of kmers *Motifs* in our DNA sample\n",
      "\n",
      "This yields our \n",
      "### Equivalent Motif Finding Problem: \n",
      "Given a collection of strings, find a collection of k-mers (one from each string) that minimizes the distance between all possible patterns and all possible collections of k-mers.\n",
      "     * Input: A collection of strings Dna and an integer k.\n",
      "     * Output: A k-mer Pattern and a collection of k-mers, one from each string in Dna, minimizing d(Pattern, Motifs) among all possible choices of Pattern and Motifs."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def d(Pattern,Motifs):\n",
      "    summation=0\n",
      "    for Motif in Motifs:\n",
      "        summation+=find_Ham(Pattern,Motif)\n",
      "    return summation            "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Equivalent Motif Finding Problem: Given a collection of strings, find a collection of k-mers (one from each string) that minimizes the distance between all possible patterns and all possible collections of k-mers.\n",
      "     Input: A collection of strings Dna and an integer k.\n",
      "     Output: A k-mer Pattern and a collection of k-mers, one from each string in Dna, minimizing\n",
      "     d(Pattern, Motifs) among all possible choices of Pattern and Motifs.\n",
      "### The Median String Problem\n",
      "\n",
      "Why doesn't this switch in approaches make things harder? Because we don't have to search **every** motif in order to minimize d(Pattern,Motifs)\n",
      "\n",
      "To further explain this- let $Motifs(Pattern,DNA)$ Be a collection of kmers that minimizes d(Pattern,Motifs) for a given Pattern and all possible kmers *Motif* in the DNA  For example, in the problem below the five colored strings represents $Motifs(\"AAA\",Dna)$\n",
      "<img src=\"img_2Motifs.png\">\n",
      "\n",
      "\n",
      "The reason why we don\u2019t need to consider all possible collections Motifs is that we can generate the strings in Motifs(Pattern, Dna) one at a time. Given a k-mer Pattern and a longer string Text, we use d(Pattern, Text) to denote the minimum Hamming distance between Pattern and any k-mer in Text,$$d(Pattern,Text)=min(HammingDistance(Pattern,Pattern'))$$ where we take the minimum Hamming Distance of all the Patterns in the Text.\n",
      "\n",
      "So, for example: \n",
      "<img src=\"img3_dexample.png\">  \n",
      "\n",
      "A k-mer in Text that achieves the minimum Hamming distance with Pattern is denoted Motif(Pattern, Text). For the above example,\n",
      "\n",
      "<img src=\"img4_Motifexample.png\"> \n",
      "\n",
      "Another definition:\n",
      "\n",
      "$$d(Pattern,DNA)= \\sum_{i=0}^{t} d(Pattern,DNA_{i})$$\n",
      "\n",
      "For example, for the strings below, $d(\"AAA\",DNA)=1+1+2+0+1=5$\n",
      "\n",
      "<img src=\"img5_dcount.png\"> \n",
      "\n",
      "Our goal is to find a k-mer Pattern that minimizes d(Pattern, Dna) over all k-mers Pattern, the same task that the Equivalent Motif Finding Problem is trying to achieve. We call such a k-mer a median string for Dna.\n",
      "\n",
      "**Median String Problem: Find a median string.**\n",
      "     Input: A collection of strings Dna and an integer k.\n",
      "     Output: A k-mer Pattern that minimizes d(Pattern, Dna) among all k-mers Pattern.\n",
      "\n",
      "Notice that finding a median string requires solving a double minimization problem. We must find a k-mer Pattern that minimizes d(Pattern, Dna), where this function is itself computed by taking a minimum over all choices of k-mers from each string in Dna."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def DistanceBetweenStrings(Pattern,DNA):\n",
      "    k=len(Pattern)\n",
      "    distance=0\n",
      "    for string in DNA:\n",
      "        HammingDistance=1*(10**10)\n",
      "        for i in range(len(string)-k):\n",
      "            if HammingDistance>find_Ham(Pattern,string[i:i+k]):\n",
      "                HammingDistance=find_Ham(Pattern,string[i:i+k])\n",
      "        distance+=HammingDistance\n",
      "    return distance\n",
      "\n",
      "def MedianString(DNA,k):\n",
      "    distance = 1*(10**100)\n",
      "    Median=\"\"\n",
      "    for Pattern in [MotMod.NumberToPattern(i,k) for i in range(4**k)] :\n",
      "        if distance > DistanceBetweenStrings(Pattern,DNA):\n",
      "            distance=DistanceBetweenStrings(Pattern,DNA)\n",
      "            Median=Pattern\n",
      "    print distance\n",
      "    return \"\".join(Median)\n",
      "\n",
      "# print MedianString(MotMod.readintextdata(\"DBPT.txt\"),6)\n",
      "\n",
      "DNA=[\"CTCGATGAGTAGGAAAGTAGTTTCACTGGGCGAACCACCCCGGCGCTAATCCTAGTGCCC\",\n",
      "\"GCAATCCTACCCGAGGCCACATATCAGTAGGAACTAGAACCACCACGGGTGGCTAGTTTC\",\n",
      "\"GGTGTTGAACCACGGGGTTAGTTTCATCTATTGTAGGAATCGGCTTCAAATCCTACACAG\"]\n",
      "\n",
      "print DistanceBetweenStrings(\"TCTGAAG\",DNA)\n",
      "print DistanceBetweenStrings(\"GTCAGCG\",DNA)\n",
      "print DistanceBetweenStrings(\"GGTTACT\",DNA)\n",
      "print DistanceBetweenStrings(\"TAGTTTC\",DNA)\n",
      "print DistanceBetweenStrings(\"GATGAGT\",DNA)\n",
      "print DistanceBetweenStrings(\"CGTGTAA\",DNA)\n",
      "print MedianString(DNA,7)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "8\n",
        "9\n",
        "6\n",
        "3\n",
        "4\n",
        "9\n",
        "0"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "AATCCTA\n"
       ]
      }
     ],
     "prompt_number": 7
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Why have we reformulated the Motif Problem?\n",
      "\n",
      "simply put: time. The running time of the BruteForceAlgorithim was $ O(n^{t} * k * t)}$ (where t is the number of strings of DNA), and the running time of the MedianStringAlgorithim was $O(4^{k} * n * t  * k}$ which is far more favorable since k is almost always less than or equal to 20, whereas t is in the thousands.\n",
      "\n",
      "The Median String Problem teaches an important lesson, which is that sometimes rethinking how a problem is formulated can lead to dramatic improvements in the runtime required to solve it. In this case, our simple observation that Score(Motifs) could just as easily be computed row-by-row as column-by-column produced the faster MEDIANSTRING algorithm.\n",
      "\n",
      "Of course, the ultimate test of a bioinformatics algorithm is how it performs in practice. Unfortunately, since MEDIANSTRING has to consider 4k k-mers, it becomes too slow for the Subtle Motif Problem, for which k = 15. We will run MEDIANSTRING with k = 13 in the hope that it will capture a substring of the correct 15-mer motif. The algorithm still requires half a day to run on our computer and returns the median string AAAAAtAGaGGGG (with distance 29). This 13-mer is not a substring of the implanted pattern AAAAAAAAGGGGGGG. but it does come close.\n",
      "\n",
      "STOP and Think: How can a slightly incorrect median string of length 13 help us find the correct median string of length 15?\n",
      "\n",
      "We have thus far assumed that the value of k is known in advance, which is not the case in practice. As a result, we are forced to run our motif finding algorithms for different values of k and then try to deduce the correct motif length. Since some regulatory motifs are rather long \u2014 later in the chapter, we will search for a biologically important motif of length 20 \u2014 MEDIANSTRING may be too slow to find them."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Using the ProfileMatrix to roll dice"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def Profile_MostProbable(Text,k,ProfileMatrix):\n",
      "    maximum=(\"\",0)\n",
      "    for i in range(len(Text)-k+1):\n",
      "        if computeprobability(Text[i:i+k],ProfileMatrix)>maximum[1]:\n",
      "            maximum=(Text[i:i+k],computeprobability(Text[i:i+k],ProfileMatrix))\n",
      "#         if \"GTCA\" in Text:\n",
      "#             print Text[i:i+k],computeprobability(Text[i:i+k],ProfileMatrix)\n",
      "    if maximum[1]==0:\n",
      "        return Text[:k]\n",
      "    else:\n",
      "        return maximum[0]\n",
      "\n",
      "        \n",
      "def computeprobability(string,ProfileMatrix):\n",
      "    string=list(string);index={\"A\":0,\"C\":1,\"G\":2,\"T\":3}\n",
      "    probability=1;\n",
      "    for i in range(len(string)):\n",
      "        probability*=ProfileMatrix[index[string[i]],i]\n",
      "    return probability\n",
      "\n",
      "def readinmatrix(filename):\n",
      "    f=open(filename,\"r\")\n",
      "    Data=f.read()\n",
      "    Data=Data.split(\"\\n\")\n",
      "    stg=[];mat=[]\n",
      "    for datum in Data:\n",
      "        moredata=datum.split(\" \")\n",
      "        for number in moredata:\n",
      "            stg.append(float(number))\n",
      "        mat.append(stg)\n",
      "        stg=[]\n",
      "    return mat\n",
      "\n",
      "# testtext=\"\"\n",
      "matrix=np.array(readinmatrix(\"matrixinput.txt\"))\n",
      "print computeprobability(\"TCGGTA\",matrix)\n",
      "# print Profile_MostProbable(testtext,12,matrix)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "0.00405\n"
       ]
      }
     ],
     "prompt_number": 10
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Now let's formalize a \"GreedyMotifSearch\" based upon our work here, and taking into account the fact we'll need to search several DNA strings"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def readintextdata(filename,option=False):\n",
      "    f=open(filename,\"r\")\n",
      "    Input=f.read()\n",
      "    Text=[]\n",
      "    if option==False: \n",
      "        Data=Input.split(\" \")\n",
      "    else:\n",
      "        Data=Input.split(\"\\n\")\n",
      "    for datum in Data:\n",
      "        Text.append(str(datum))\n",
      "    f.close()\n",
      "    return Text"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 6
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def GreedyMotifSearch(DNA,k,t):\n",
      "    BestMotifs=[]\n",
      "    for strand in DNA:\n",
      "        BestMotifs.append(strand[0:k])\n",
      "    basestrand=DNA[0]\n",
      "    otherstrands=[DNA[i] for i in range(1,len(DNA))]\n",
      "    for kmer in [basestrand[i:i+k] for i in range(len(basestrand)-k)]:\n",
      "        Motifs=[];Motifs.append(kmer)\n",
      "        for strand in otherstrands:\n",
      "            ProfileMatrix=ProfileMat(CountMatPseudo(MotifMat(Motifs)))\n",
      "            nextMotif=Profile_MostProbable(strand,k,ProfileMatrix)\n",
      "            Motifs.append(nextMotif)\n",
      "        if MotifScore(Motifs,k)<MotifScore(BestMotifs,k):\n",
      "            BestMotifs=Motifs\n",
      "    return BestMotifs\n",
      "\n",
      "def MotifScore(Motifs,k):\n",
      "    consensus=Get_Consensus(ProfileMat(CountMat(MotifMat(Motifs))))\n",
      "    score=0\n",
      "    for motif in Motifs:\n",
      "        score+=find_Ham(motif,consensus)\n",
      "    return score\n",
      "\n",
      "# answersattempt=GreedyMotifSearch(readintextdata(\"GTI.txt\",True),12,25)\n",
      "# for answer in answersattempt:\n",
      "#     print answer\n",
      "# for answer in answers:\n",
      "#     answer.replace(\"\",\" \")\n",
      "# for answer in answersattempt:\n",
      "#     if answer not in answers:\n",
      "#         print answer\n",
      "# print \"And now for the other side\"\n",
      "# for answer in answers:\n",
      "#     if answer not in answersattempt:\n",
      "#         print answer"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 7
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Randomized Motif Search\n",
      "\n",
      "Randomized Algorithms can be surprisingly powerful tools to efficiently find answers to problems that cannot be easily deterministically derived "
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import random\n",
      "def Motifs_from_Profile(Profile,DNA):\n",
      "    k=len(Profile[0])\n",
      "    Motifs=[]\n",
      "    for strand in DNA:\n",
      "        Motifs.append(Profile_MostProbable(strand,k,Profile))\n",
      "    return Motifs\n",
      "\n",
      "def random_kmer(string,k):\n",
      "    x=len(string)\n",
      "    while len(string)-x<k:\n",
      "        x=random.randrange(0,len(string))\n",
      "    return string[x:x+k]\n",
      "\n",
      "def initialize_randomMotifs(k,DNA):\n",
      "    Motifs=[]\n",
      "    for string in DNA:\n",
      "        Motifs.append(random_kmer(string,k))\n",
      "    return Motifs\n",
      "\n",
      "def RandomizedMotifSearch(DNA,k,t):\n",
      "    i=0;j=0;a=0;switch=True\n",
      "    while i<1000:\n",
      "        random.seed(a);a+=1\n",
      "        Motifs=initialize_randomMotifs(k,DNA)\n",
      "        if i==0:\n",
      "            BestMotifs=Motifs\n",
      "        while 1==1:\n",
      "            ProfileMatrix=ProfileMat(CountMatPseudo(MotifMat(Motifs)))\n",
      "            Motifs=Motifs_from_Profile(ProfileMatrix,DNA)\n",
      "            if MotifScore(Motifs,k)<MotifScore(BestMotifs,k):\n",
      "                BestMotifs=Motifs\n",
      "            else:\n",
      "                break\n",
      "        i+=1\n",
      "    return BestMotifs\n",
      "\n",
      "\n",
      "\n",
      "# answers=RandomizedMotifSearch(readintextdata(\"RMS.txt\",True),15,20)\n",
      "# print \"\\n\".join(answers)\n",
      "# print MotifScore(answers,15) # Works!"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 8
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Gibbs Sampling:\n",
      "\n",
      "A more cautious method of finding the best motifs, using a more \"advanced\" Random Number generator"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def Select_Pseudo_RandMotif(X,ProfileMatrix,k):\n",
      "    # from all the kmers in the \"special strand\"\n",
      "    kmers=[X[i:i+k] for i in range(len(X)-k)]\n",
      "    #compute the probability that that kmer is a motif dictated by the profile matrix\n",
      "    probabilities=[computeprobability(kmer,ProfileMatrix) for kmer in kmers]\n",
      "    BD=[x/sum(probabilities) for x in probabilities]\n",
      "    randroll=random.random();summation=float(0)\n",
      "    for i in range(len(BD)):\n",
      "        summation+=BD[i]\n",
      "        if randroll<summation:\n",
      "            randomkmer=i\n",
      "            break\n",
      "    return kmers[randomkmer]\n",
      "\n",
      "def GibbsSampler(DNA,k,t,N):\n",
      "    Motifs=initialize_randomMotifs(k,DNA) # Select a random kmer motif in each DNA string as starter set\n",
      "    BestMotifs=list(Motifs);switch=0\n",
      "#     print \"initial\" + str(MotifScore(BestMotifs,k))\n",
      "    for j in range(N):#Run N iterations\n",
      "        X=random.randrange(0,len(DNA))\n",
      "        ProfileMatrix=ProfileMat(CountMatPseudo(MotifMat([Motifs[i] for i in range(len(Motifs)) if i!=X])))\n",
      "        Motifs[X]=Select_Pseudo_RandMotif(DNA[X],ProfileMatrix,k)\n",
      "        if MotifScore(Motifs,k)<MotifScore(BestMotifs,k):\n",
      "            BestMotifs=list(Motifs)\n",
      "            switch+=1\n",
      "    print BestMotifs,MotifScore(BestMotifs,k)\n",
      "    return BestMotifs\n",
      "\n",
      "# testing=[\"CGCCCCTCTCGGGGGTGTTCAGTAAACGGCCA\",\n",
      "#          \"GGGCGAGGTATGTGTAAGTGCCAAGGTGCCAG\",\n",
      "#          \"TAGTACCGAGACCGAAAGAAGTATACAGGCGT\", \n",
      "#          \"TAGATCAAGTTTCAGGTGCACGTCGGTGAACC\",\n",
      "#          \"AATCCACCAGCTCCACGTGCAATGTTGGCCTA\"]\n",
      "# answersone=set(GibbsSampler(testing,8,5,1000))\n",
      "# answers=set([\"TCTCGGGG\",\"CCAAGGTG\",\"TACAGGCG\",\"TTCAGGTG\",\"TCCACGTG\"])\n",
      "# print MotifScore(answers,8)\n",
      "\n",
      "print \"\\n\".join(GibbsSampler(readintextdata(\"Gibbs.txt\",True),15,20,2000))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "['ACATTCGCACCGCAG', 'AGTAGAACACCTGAG', 'AGTCAAACACCTGAG', 'AGTTTCACTATTGAG', 'AAGGTCACACCTGAG', 'AGTTTCTGTCCTGAG', 'TTTTTCACACCTGAT', 'TTATTCACACCTGAG', 'AGTTTCACAAGCGAG', 'AGTTTTGAACCTGAG', 'AGCGACACACCTGAG', 'AGTTTCTGTCCTGAG', 'AGTTTCACACCATGG', 'AGTTTAGTACCTGAG', 'AGTTTCACACCTCCC', 'AGTTTCACACGCCAG', 'AGTTTCAGCACTGAG', 'CGTTTCACACCTGTC', 'AGTTCTGCACCTGAG', 'AGTTGTTCACCTGAG'] 62\n",
        "ACATTCGCACCGCAG\n",
        "AGTAGAACACCTGAG\n",
        "AGTCAAACACCTGAG\n",
        "AGTTTCACTATTGAG\n",
        "AAGGTCACACCTGAG\n",
        "AGTTTCTGTCCTGAG\n",
        "TTTTTCACACCTGAT\n",
        "TTATTCACACCTGAG\n",
        "AGTTTCACAAGCGAG\n",
        "AGTTTTGAACCTGAG\n",
        "AGCGACACACCTGAG\n",
        "AGTTTCTGTCCTGAG\n",
        "AGTTTCACACCATGG\n",
        "AGTTTAGTACCTGAG\n",
        "AGTTTCACACCTCCC\n",
        "AGTTTCACACGCCAG\n",
        "AGTTTCAGCACTGAG\n",
        "CGTTTCACACCTGTC\n",
        "AGTTCTGCACCTGAG\n",
        "AGTTGTTCACCTGAG\n"
       ]
      }
     ],
     "prompt_number": 48
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}