{
 "metadata": {
  "name": "",
  "signature": "sha256:6ccea312020949e4a95f9caedba09771e4e0be58463178147ed5c99354b94f37"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Sequencing Antibiotics\n",
      "-------\n",
      "\n",
      "#RNA-->AA\n",
      "We first begin by translating RNA script into AA\n"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from Bio.Seq import Seq\n",
      "from Bio.Alphabet import IUPAC,generic_dna\n",
      "\n",
      "# f=open(\"abinput_one.txt\",\"r\")\n",
      "# string=f.read()\n",
      "# o=open(\"aboutput_one.txt\",\"w\")\n",
      "\n",
      "# o.write(str(Seq(string,generic_rna).translate(to_stop=True)))\n",
      "# f.close()\n",
      "# o.close()\n",
      "\n",
      "              "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Reverse Translate!"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "tyrocine = Seq(\"VKLFPWFNQY\", IUPAC.protein)\n",
      "\n",
      "# Example code to follow below\n",
      "#my_protein = Seq('AVMGRWKGGRAAG', IUPAC.protein)\n",
      "# standard_translator.back_translate(my_protein)\n",
      "# Seq('GCTGTTATGGGTCGTTGGAAGGGTGGTCGTGCTGCTGGT', IUPACUnambiguousDNA())"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "NameError",
       "evalue": "name 'standard_translator' is not defined",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
        "\u001b[0;32m<ipython-input-13-1ed3577cd951>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mtyrocine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSeq\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"VKLFPWFNQY\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mIUPAC\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprotein\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mstandard_translator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mback_translate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtyrocine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;31m# Example code to follow below\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m#my_protein = Seq('AVMGRWKGGRAAG', IUPAC.protein)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
        "\u001b[0;31mNameError\u001b[0m: name 'standard_translator' is not defined"
       ]
      }
     ],
     "prompt_number": 13
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "We say that a DNA string Pattern encodes an amino acid string Peptide if the RNA string transcribed from either Pattern or its reverse complement Pattern translates into Peptide. For example, the DNA string GAAACT is transcribed into GAAACU and translated into ET. The reverse complement of this DNA string, AGTTTC, is transcribed into AGUUUC and translated into SF. Thus, GAAACT encodes both ET and SF.\n",
      "\n",
      "\n",
      "Peptide Encoding Problem: Find substrings of a genome encoding a given amino acid sequence.\n",
      "     Input: A DNA string Text, an amino acid string Peptide, and the array GeneticCode.\n",
      "     Output: All substrings of Text encoding Peptide (if any such substrings exist)."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from string import ascii_uppercase\n",
      "\n",
      "# thinking of ways to speed up function- instead of comparing translated dna or dna complement to peptide, reverse translate all the possible\n",
      "# combinations of Peptide and run those against the text codons and text codons complement. May or may not speed up algorithm slightly\n",
      "def find_peptide(Text,Peptide):\n",
      "    Text=Seq(Text,IUPAC.ambiguous_dna)\n",
      "    textlength=len(str(Text))\n",
      "    num_nucleotides=len(Peptide)*3\n",
      "    sequencesofinterest=[]\n",
      "    i=0\n",
      "    while i<textlength-num_nucleotides:\n",
      "        codonsofinterest=Text[i:i+num_nucleotides]\n",
      "        complement=codonsofinterest.reverse_complement()\n",
      "        i+=1\n",
      "        try:\n",
      "            if str(codonsofinterest.translate())==Peptide or str(complement.translate())==Peptide:\n",
      "                sequencesofinterest.append(str(codonsofinterest))\n",
      "        except:\n",
      "            print i\n",
      "    return sequencesofinterest\n",
      "\n",
      "# f=open(\"peptidefindinput.txt\",\"r\") # Used to answer the exercise\n",
      "# Codons=str(f.readline())\n",
      "# AA=str(f.readline())\n",
      "# f.close()\n",
      "# f=open(\"B_brevis.txt\",\"r\")\n",
      "# Codons=str(f.read())\n",
      "# Codons=Codons.strip()\n",
      "# AA=\"VKLFPWFNQY\"\n",
      "# f.close()\n",
      "# String=\"\"\n",
      "# for i in Codons:\n",
      "#     if i in ascii_uppercase:\n",
      "#         String+=str(i)\n",
      "\n",
      "# answer = find_peptide(String,AA)\n",
      "# print len(answer)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Finding Bacterially sequenced proteins\n",
      "\n",
      "Because tyrocine (our antibioitic of interest here) is a cyclic peptide, there are a great many number of possible peptides\n",
      "we would look for, but none of them are found in sequence (right next to each other) within the bacillus brevis genome"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Instead of using genome sequencing, let's use mass spectrometry"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from Bio.SeqUtils.ProtParam import ProteinAnalysis\n",
      "import itertools\n",
      "## find number of possible number of subpeptides from a peptide of length n\n",
      "\n",
      "def num_subpeptides(n):\n",
      "    return (n-1)*n\n",
      "# return list of all masses of possible subpeptides\n",
      "\n",
      "\n",
      "def find_mass(Peptide):\n",
      "    AAs={\"G\": 57,\"A\": 71,\"S\": 87, \"P\": 97, \"V\": 99, \"T\": 101, \"C\": 103, \"I\": 113, \"L\": 113, \"N\": 114, \"D\": 115, \"K\": 128, \"Q\": 128, \"E\": 129, \"M\": 131, \"H\": 137, \"F\": 147, \"R\": 156, \"Y\": 163, \"W\": 186}\n",
      "    mass=0\n",
      "    for i in Peptide:\n",
      "        mass+=AAs[i]\n",
      "    return mass\n",
      "\n",
      "def Cyclospectrum(Peptide):\n",
      "    masses=[0,find_mass(Peptide)];i=len(Peptide)\n",
      "    while i!=1:\n",
      "        i-=1\n",
      "        for z in itertools.combinations(Peptide,i):\n",
      "            pep=\"\".join(z)\n",
      "            masses.append(find_mass(pep))\n",
      "    return masses\n",
      "\n",
      "## still not sure how they're generating all the possible combinations of subpeptides\n",
      "# Peptide=\"IAQMLFYCKVATN\"\n",
      "Peptide=\"PNGPSQGMPSRAG\"\n",
      "answer= Cyclospectrum(Peptide)\n",
      "answer.sort()\n",
      "f=open(\"cyclospectrumoutput.txt\",\"w\")\n",
      "for i in answer:\n",
      "    f.write(str(i)+\" \")\n",
      "f.close()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "The theoretical spectrum of a cyclic peptide *Peptide* denoted *Cyclospectrum(Peptide)* is the collection of all of the masses of its subpeptides, in addition to the mass 9 and the mass of the entire peptide with masses ordered from smallest to largest. We will assume that the theoretical spectrum can contain duplicate elements as is the case for NQEL where NQ and EL have the same mass.\n",
      "\n",
      "Above, is my first attempt (solo attempt) to determine the CycloSpectrum using a combination generator. Unfortunately, this didn't work, so I followed the book's instructions (below) to determine the Linear and Cyclic Spectrums, which *did* work."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "AAs=[\"G\",\"A\",\"S\",\"P\",\"V\",\"T\",\"C\",\"I\",\"L\",\"N\",\"D\",\"K\",\"Q\",\"E\",\"M\",\"H\",\"F\",\"R\",\"Y\",\"W\"]\n",
      "AAs_mass=[57,71,87,97,99,101,103,113,113,114,115,128,128,129,131,137,147,156,163,186]\n",
      "\n",
      "def LinearSpectrum(Peptide,AminoAcid,AminoAcidMass):\n",
      "    PrefixMasses=[0]\n",
      "    for i in range(0,len(Peptide)):\n",
      "        for j in range(0,20):\n",
      "            if AminoAcid[j]==Peptide[i]:\n",
      "                PrefixMasses.append(PrefixMasses[i]+AminoAcidMass[j])\n",
      "    LinearSpec=[0]\n",
      "    for i in range(0,len(Peptide)):\n",
      "        for j in range(i+1,len(Peptide)+1):\n",
      "            LinearSpec.append(PrefixMasses[j]-PrefixMasses[i])\n",
      "    LinearSpec.sort()\n",
      "    return LinearSpec\n",
      "\n",
      "f=open(\"spectrumoutput.txt\",\"w\")\n",
      "for i in LinearSpectrum(\"PMGIEHGLNAASTGAQYSLAKPYDPTLVGGSRTVHERCKQSHMAAECF\",AAs,AAs_mass):\n",
      "    f.write(str(i)+ \" \")\n",
      "f.close()\n",
      "\n",
      "def CyclicSpectrum(Peptide,AminoAcid,AminoAcidMass):\n",
      "    PrefixMasses=[0]\n",
      "    for i in range(0,len(Peptide)):\n",
      "        for j in range(0,20):\n",
      "            if AminoAcid[j]==Peptide[i]:\n",
      "                PrefixMasses.append(PrefixMasses[i]+AminoAcidMass[j])\n",
      "    peptideMass= PrefixMasses[len(Peptide)]\n",
      "    CyclicSpec=[0]\n",
      "    for i in range(0,len(Peptide)):\n",
      "        for j in range(i+1,len(Peptide)+1):\n",
      "            CyclicSpec.append(PrefixMasses[j]-PrefixMasses[i])\n",
      "            if i>0 and j<len(Peptide):\n",
      "                CyclicSpec.append(peptideMass-(PrefixMasses[j]-PrefixMasses[i]))\n",
      "    CyclicSpec.sort()\n",
      "    return CyclicSpec\n",
      "\n",
      "z=open(\"cyclicspectrumoutput.txt\",\"w\")\n",
      "for i in CyclicSpectrum(\"TYVERRKQPPEHPYY\",AAs,AAs_mass):\n",
      "    z.write(str(i)+\" \")\n",
      "z.close()\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "####Well that's great but...\n",
      "\n",
      "so it's great that we can find theoretical spectrums of a known peptide...but that's kinda easy. We want to find an unknown peptide from its *experimental* spectrum. Let's start by assuming that a biologist is lucky enough to generate an ideal experimental spectrum, which is one coinciding with the peptide's theoreteical spectrum.\n",
      "\n",
      "One method you might employ to find a peptide by its spectrum, is to construct (by combination generator) every possible peptide of that weight. This is a **brute force algorithm**, an algorithim that searches every possible candidate solution in order to find the true solution.\n",
      "\n",
      "\n",
      "In their words: \n",
      "\n",
      "####Cyclopeptide Sequencing Problem:\n",
      "*Given an ideal experimental spectrum, find a cyclic peptide whose theoretical spectrum matches the experimental spectrum.*\n",
      "**Input:** A collection of (possibly repeated integers *Spectrum* corresponding to an ideal experimental spectrum.\n",
      "**Output: ** An amino acid string *Peptide* such that the *Cyclospectrum(Peptide)= Spectrum* if such a string exists\n",
      "\n"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Let's demonstrate the infeasibility of this brute force approach by computing the number of peptidse of a given total mass.\n",
      "**Input:** An integer m\n",
      "**Output:** (integer)The number of linear peptides having integer mass m\n"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "##Work on this in a bit-with dynamic programming help from http://www.geeksforgeeks.org/dynamic-programming-set-7-coin-change/"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "## finding the number of subpeptides for a linear peptide\n",
      "\n",
      "def find_linearsubpeptides(n):\n",
      "    sum=0\n",
      "    for i in range(n+1):\n",
      "        if i==0 or i==n:\n",
      "            sum+=1\n",
      "        else:\n",
      "            sum+=(n-(i-1))\n",
      "    return sum\n",
      "print find_linearsubpeptides(13278)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "88159282\n"
       ]
      }
     ],
     "prompt_number": 23
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#Reference: \n",
      "AAs=[\"G\",\"A\",\"S\",\"P\",\"V\",\"T\",\"C\",\"I\",\"L\",\"N\",\"D\",\"K\",\"Q\",\"E\",\"M\",\"H\",\"F\",\"R\",\"Y\",\"W\"]\n",
      "AAs_mass=[57,71,87,97,99,101,103,113,113,114,115,128,128,129,131,137,147,156,163,186]\n",
      "##Branch and Bound Algorithim Implimentation\n",
      "def CycloPeptideSequencing(Spectrum,AAs_mass,AAs):\n",
      "    Peptides=[];initializePeptides(Peptides,Spectrum,AAs_mass,AAs) ## Initialize Peptides List with potential peptides based upon theoretical spectrum\n",
      "    candidates=[]\n",
      "    droplist=[]\n",
      "    weights=[]\n",
      "    while len(Peptides)!=0:\n",
      "        Peptides=Expand(Peptides,AAs_mass,AAs) #Expand Peptides beyond initialization, creating appropriate kmers for testing\n",
      "        for peptide in Peptides:\n",
      "            if Spectrum_Consistent(CyclicSpectrum(peptide,AAs,AAs_mass),Spectrum)==False: # Trim peptide list based upon consistency\n",
      "                droplist.append(peptide)\n",
      "            else:\n",
      "                if find_mass(peptide)==Spectrum[len(Spectrum)-1]: #if consistent check for complete match\n",
      "                    for AminoAcid in peptide: # if its a complete match, make ordered list of peptide weights, append them to final list\n",
      "                        weights.append(find_mass(str(AminoAcid)))\n",
      "                    candidates.append(list(weights))\n",
      "                    droplist.append(peptide)\n",
      "                    weights[:]=[]\n",
      "        for pep in droplist:\n",
      "            Peptides.remove(pep)\n",
      "        droplist[:]=[]\n",
      "        \n",
      "    uniquecandidates=findUnique(candidates)\n",
      "    return uniquecandidates\n",
      "\n",
      "#returns unique list elements from input list\n",
      "def findUnique(listofinterest):\n",
      "    unique=[]\n",
      "    for i in listofinterest:\n",
      "        if i not in unique:\n",
      "            unique.append(i)\n",
      "    return unique\n",
      "\n",
      "# determines all singular amino acids whose masses are found within experimental spectrum input \n",
      "#Input: Empty list of Peptides, Experimental Spectrum, Reference Lists\n",
      "#Output: Emtpy \n",
      "def initializePeptides(Peptides,ExperimentalSpectrum,AAs_mass,AAs):\n",
      "    for i in range(len(ExperimentalSpectrum)):\n",
      "        for j in range(20):\n",
      "            if AAs_mass[j]==ExperimentalSpectrum[i]:\n",
      "                Peptides.append(AAs[j])\n",
      "    return Peptides\n",
      "\n",
      "#Adds on each Amino Acid to a prexisting list \"Peptides\", returning that list, also removes initial peptides from input list\n",
      "#Input:Peptides list, Reference Arrays\n",
      "#Output: \"Expanded\" list of Peptides\n",
      "def Expand(Peptides,AAs_mass,AAs):\n",
      "    Expanders=list(Peptides)\n",
      "    for AminoAcid in Expanders:\n",
      "        for i in AAs:\n",
      "            Peptides.append(str(AminoAcid+i))\n",
      "    for AminoAcid in Expanders:\n",
      "        if AminoAcid in Peptides:\n",
      "            Peptides.remove(AminoAcid)\n",
      "    return Peptides\n",
      "\n",
      "#Input= two lists (Spectra)\n",
      "#Output= Boolean denoting whether every value in ProposedPeptide Spec can be found in the Experimental\n",
      "def Spectrum_Consistent(ProposedPeptideSpec,ExperimentalPeptideSpec):\n",
      "    for mass in ProposedPeptideSpec:\n",
      "        if mass not in ExperimentalPeptideSpec:\n",
      "            return False\n",
      "    return True\n",
      "\n",
      "def readindata(filename):\n",
      "    f=open(filename,\"r\")#read in quiz data\n",
      "    QuizInput=f.read()\n",
      "    QuizDataInput=[]\n",
      "    Data=QuizInput.split(\" \")\n",
      "    for datum in Data:\n",
      "        QuizDataInput.append(int(datum))\n",
      "    f.close()\n",
      "    return QuizDataInput\n",
      "txtfile=\"branchboundinput.txt\"\n",
      "\n",
      "\n",
      "print CycloPeptideSequencing(readindata(txtfile),AAs_mass,AAs)\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SampleInput=[0, 113, 128, 186, 241, 299, 314, 427] \n",
      "answers=[[186,128,113], [186,113,128], [128,186,113], [128,113,186], [113,186,128], [113,128,186]]\n",
      "for weight in  CycloPeptideSequencing(SampleInput,AAs_mass,AAs):\n",
      "    print weight\n",
      "\n",
      "\n",
      "##Experimentation!\n",
      "# testing=[\"I\",\"L\",\"K\",\"Q\",\"W\"]\n",
      "# #expand\n",
      "# stg=Expand(testing,AAs_mass,AAs)\n",
      "# badpeplist=[]\n",
      "# monkey=False\n",
      "# #trim\n",
      "# for pep in stg:\n",
      "#     if Spectrum_Consistent(CyclicSpectrum(pep,AAs,AAs_mass),SampleInput)==False:\n",
      "#         badpeplist.append(pep)\n",
      "# for pep in badpeplist: #actual trimming\n",
      "#     stg.remove(pep)\n",
      "# badpeplist[:]=[]\n",
      "# #expand\n",
      "# stg=Expand(stg,AAs_mass,AAs)\n",
      "# for pep in stg:\n",
      "#     if Spectrum_Consistent(CyclicSpectrum(pep,AAs,AAs_mass),SampleInput)==False:\n",
      "#         badpeplist.append(pep)\n",
      "#     elif CyclicSpectrum(pep,AAs,AAs_mass)==SampleInput:\n",
      "#             print pep, CyclicSpectrum(pep,AAs,AAs_mass), find_mass(pep[0]),find_mass(pep[1]), find_mass(pep[2])\n",
      "# for pep in badpeplist: #trimming\n",
      "#     stg.remove(pep)\n",
      "# badpeplist[:]=[]\n",
      "# #expand\n",
      "# stg=Expand(stg,AAs_mass,AAs) #expanding\n",
      "# for pep in stg:\n",
      "#     if Spectrum_Consistent(CyclicSpectrum(pep,AAs,AAs_mass),SampleInput)==False:\n",
      "#         badpeplist.append(pep)\n",
      "# for pep in badpeplist: # trimming\n",
      "#     stg.remove(pep)\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "[]\n",
        "[113, 128, 186]\n",
        "[113, 186, 128]\n",
        "[128, 113, 186]\n",
        "[128, 186, 113]\n",
        "[186, 113, 128]\n",
        "[186, 128, 113]\n"
       ]
      }
     ],
     "prompt_number": 143
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}